### PR TITLE
Expanded coil documentation

### PR DESCRIPTION
**Pull request by rogeriojorge**
In this PR, the novel features of stage2 optimization in SIMSOPT are documented.
Namely it is added documentation on:
- Stochastic optimization
- Finite beta optimization
- Finite build optimization
The documentation follows the examples provided in the examples folder.

### CODE REVIEW

File: docs/source/example_coils.rst [hunk 1]
```diff
@@ -302,3 +346,263 @@ for coil ``j`` can be obtained from ``coils[j].curve.x``, where the
...
+
+Stochastic Optimization
+---------------------------
+
+In this example we solve a stochastic version of
+the :ref:`first example here <_simplest_stage2>`. As before,
+the goal is to find coils that generate a specific target
+normal field on a given surface. As we are still considering a vacuum
+field the target is just zero.
+The target equilibrium is the precise QA configuration of arXiv:2108.03711.
```
Comment (florianwechsung) [inline docs/source/example_coils.rst:358]: Could you go through the file and replace all the arxiv references with the doi of the published paper?

File: docs/source/example_coils.rst [hunk 2]
```diff
@@ -302,3 +346,263 @@ for coil ``j`` can be obtained from ``coils[j].curve.x``, where the
...
+points of the curve. If ``nintervals="partial"``, then the argument is as follows:
+A curve in 3d space is defined uniquely by an initial point, an initial
+direction, and the arclength, curvature, and torsion along the curve. For a
+:mod:`simsopt.geo.curvexyzfourier.CurveXYZFourier`, the intuition is now as
+follows: assuming that the curve has order :math:`p`, that means we have
+:math:`3*(2p+1)` degrees of freedom in total. Assuming that three each are
+required for both the initial position and direction, :math:`6p-3` are left
+over for curvature, torsion, and arclength. We want to fix the arclength,
+so we can afford :math:`2p-1` constraints, which corresponds to
+:math:`L=2p`.
```
Comment (florianwechsung) [inline docs/source/example_coils.rst:398]: I think this paragraph is overkill. It's very technical, so how about just keeping the first sentence and then adding a link to the docs where this is explained?

**Review by florianwechsung — COMMENTED** [review]
Thanks @rogeriojorge , great stuff! Made some edits my self and left a couple requests. Otherwise this all looks good to me.

File: docs/source/example_coils.rst (hunk 1 above)
Comment (rogeriojorge) [reply docs/source/example_coils.rst]: Done

File: docs/source/example_coils.rst (hunk 2 above)
Comment (rogeriojorge) [reply docs/source/example_coils.rst]: Done

**Comment by rogeriojorge** [comment]
Thanks for the review @florianwechsung. The suggestions have been applied.

**Review by florianwechsung — APPROVED** [review]
Looks good. I'll merge once all the ticks are green.

**Pull request merged**